### ISSUE TITLE
Failed to `step` after Broken pipe error

### ISSUE THREAD

**Issue by keisuke-nakata** [minerllabs/minerl#79]
Occasionally MineRL fails with `Broken pipe` error and never reconnect.

Here is my console log:

```
INFO     - 2019-06-19 20:57:18,227 - [chainerrl.experiments.train_agent train_agent 68] statistics:[('average_q', 0.3040304352121823), ('average_loss', 3.83408982771332), ('n_updates', 35494)]
ERROR    - 2019-06-19 20:58:09,334 - [minerl.env.malmo log_to_file 344] [20:58:09] [Client thread/INFO]: [STDOUT]: CLIENT request state: ERROR_TIMED_OUT_WAITING_FOR_EPISODE_PAUSE
ERROR    - 2019-06-19 20:58:09,335 - [minerl.env.malmo log_to_file 344] [20:58:09] [Client thread/INFO]: [STDOUT]: CLIENT enter state: ERROR_TIMED_OUT_WAITING_FOR_EPISODE_PAUSE
ERROR    - 2019-06-19 21:01:19,352 - [minerl.env.core _start_up 403] Failed to reset (timeout), trying again!
ERROR    - 2019-06-19 21:01:48,329 - [minerl.env.malmo log_to_file 344] [21:01:48] [EnvServerSocketHandler/INFO]: [STDERR]: java.net.SocketException: Broken pipe (Write failed)
ERROR    - 2019-06-19 21:01:48,330 - [minerl.env.malmo log_to_file 344] [21:01:48] [EnvServerSocketHandler/INFO]: [STDERR]: 	at java.net.SocketOutputStream.socketWrite0(Native Method)
ERROR    - 2019-06-19 21:01:48,330 - [minerl.env.malmo log_to_file 344] [21:01:48] [EnvServerSocketHandler/INFO]: [STDERR]: 	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:111)
ERROR    - 2019-06-19 21:01:48,330 - [minerl.env.malmo log_to_file 344] [21:01:48] [EnvServerSocketHandler/INFO]: [STDERR]: 	at java.net.SocketOutputStream.write(SocketOutputStream.java:134)
ERROR    - 2019-06-19 21:01:48,330 - [minerl.env.malmo log_to_file 344] [21:01:48] [EnvServerSocketHandler/INFO]: [STDERR]: 	at java.io.DataOutputStream.writeInt(DataOutputStream.java:198)
ERROR    - 2019-06-19 21:01:48,330 - [minerl.env.malmo log_to_file 344] [21:01:48] [EnvServerSocketHandler/INFO]: [STDERR]: 	at com.microsoft.Malmo.Client.MalmoEnvServer.peek(MalmoEnvServer.java:591)
ERROR    - 2019-06-19 21:01:48,330 - [minerl.env.malmo log_to_file 344] [21:01:48] [EnvServerSocketHandler/INFO]: [STDERR]: 	at com.microsoft.Malmo.Client.MalmoEnvServer.access$700(MalmoEnvServer.java:51)
ERROR    - 2019-06-19 21:01:48,330 - [minerl.env.malmo log_to_file 344] [21:01:48] [EnvServerSocketHandler/INFO]: [STDERR]: 	at com.microsoft.Malmo.Client.MalmoEnvServer$1.run(MalmoEnvServer.java:166)
ERROR    - 2019-06-19 21:01:48,330 - [minerl.env.malmo log_to_file 344] [21:01:48] [EnvServerSocketHandler/INFO]: [STDOUT]: [ERROR] MalmoEnv socket error: java.net.SocketException: Broken pipe (Write failed) (can be on disconnect)
INFO     - 2019-06-19 21:01:48,330 - [minerl.env.malmo log_to_file 346] [21:01:48] [EnvServerSocketHandler/INFO]: [STDOUT]: [LOGTOPY] Want to quit on disconnect.
INFO     - 2019-06-19 21:01:50,522 - [chainerrl.experiments.train_agent save_agent 274] Saved the agent to results/MineRLTreechop-v0/rainbow_reset_rc300000_rs5000_gsoff_fs16/20190619T063940.084949/146976_except
ERROR    - 2019-06-19 21:01:50,522 - [__main__ main 257] execution failed.
Traceback (most recent call last):
  File "dqn_family.py", line 255, in main
    _main(args)
  File "dqn_family.py", line 383, in _main
    eval_env=eval_env,
  File "/usr/local/lib/python3.5/dist-packages/chainerrl/experiments/train_agent.py", line 163, in train_agent_with_evaluation
    logger=logger)
  File "/usr/local/lib/python3.5/dist-packages/chainerrl/experiments/train_agent.py", line 54, in train_agent
    obs, r, done, info = env.step(action)
  File "/usr/local/lib/python3.5/dist-packages/gym/core.py", line 282, in step
    return self.env.step(self.action(action))
  File "/usr/local/lib/python3.5/dist-packages/gym/core.py", line 258, in step
    observation, reward, done, info = self.env.step(action)
  File "/usr/local/lib/python3.5/dist-packages/gym/core.py", line 258, in step
    observation, reward, done, info = self.env.step(action)
  File "/usr/local/lib/python3.5/dist-packages/gym/core.py", line 258, in step
    observation, reward, done, info = self.env.step(action)
  File "/mnt/vol10/nakata/home/minerl2019/minerl-chainerrl/baselines/env_wrappers.py", line 37, in step
    obs, reward, done, info = self.env.step(action)
  File "/usr/local/lib/python3.5/dist-packages/gym/core.py", line 224, in step
    return self.env.step(action)
  File "/usr/local/lib/python3.5/dist-packages/chainerrl/wrappers/continuing_time_limit.py", line 38, in step
    observation, reward, done, info = self.env.step(action)
  File "/tmp/minerl2019/minerl/minerl/env/core.py", line 496, in step
    "Attempted to step an environment with done=True")
RuntimeError: Attempted to step an environment with done=True
ERROR    - 2019-06-19 21:01:50,550 - [minerl.env.malmo log_to_file 344] [21:01:50] [EnvServerSocketHandler/INFO]: [STDERR]: java.io.EOFException
ERROR    - 2019-06-19 21:01:50,550 - [minerl.env.malmo log_to_file 344] [21:01:50] [EnvServerSocketHandler/INFO]: [STDERR]: 	at java.io.DataInputStream.readInt(DataInputStream.java:392)
ERROR    - 2019-06-19 21:01:50,551 - [minerl.env.malmo log_to_file 344] [21:01:50] [EnvServerSocketHandler/INFO]: [STDERR]: 	at com.microsoft.Malmo.Client.MalmoEnvServer$1.run(MalmoEnvServer.java:143)
ERROR    - 2019-06-19 21:01:50,551 - [minerl.env.malmo log_to_file 344] [21:01:50] [EnvServerSocketHandler/INFO]: [STDERR]: java.io.EOFException
ERROR    - 2019-06-19 21:01:50,551 - [minerl.env.malmo log_to_file 344] [21:01:50] [EnvServerSocketHandler/INFO]: [STDERR]: 	at java.io.DataInputStream.readInt(DataInputStream.java:392)
ERROR    - 2019-06-19 21:01:50,551 - [minerl.env.malmo log_to_file 344] [21:01:50] [EnvServerSocketHandler/INFO]: [STDERR]: 	at com.microsoft.Malmo.Client.MalmoEnvServer$1.run(MalmoEnvServer.java:143)
ERROR    - 2019-06-19 21:01:50,551 - [minerl.env.malmo log_to_file 344] [21:01:50] [EnvServerSocketHandler/INFO]: [STDOUT]: [ERROR] MalmoEnv socket error: java.io.EOFException (can be on disconnect)
ERROR    - 2019-06-19 21:01:50,552 - [minerl.env.malmo log_to_file 344] [21:01:50] [EnvServerSocketHandler/INFO]: [STDOUT]: [ERROR] MalmoEnv socket error: java.io.EOFException (can be on disconnect)
INFO     - 2019-06-19 21:01:50,552 - [minerl.env.malmo log_to_file 346] [21:01:50] [EnvServerSocketHandler/INFO]: [STDOUT]: [LOGTOPY] Want to quit on disconnect.
INFO     - 2019-06-19 21:01:53,556 - [minerl.env.malmo on_terminate 200] Minecraft process psutil.Process(pid=294, status='terminated') terminated with exit code 0
```

And 
[This](https://github.com/minerllabs/minerl/files/3313456/minecraft_proc_9000.log) is MineRL-side log.

It seems that the state of MineRL becomes strange, because Gym says `RuntimeError: Attempted to step an environment with done=True`, which never happen.

**Comment by keisuke-nakata** [comment]
Python 3.5.2, minerl 0.1.10 (commit 03982a3c245800e77c3b7ceaa22dc899bfaccd22)

**Comment by MadcowD** [comment]
Does this happen with `minerl` v0.1.11

**Comment by MadcowD** [comment]
Does this happen in the middle of a `step` or when the episode is being reset?

**Comment by MadcowD** [comment]
If it happens during step, then it should set done = True and you need to reset the environment.

**Comment by keisuke-nakata** [comment]
This still happens for MineRL==0.1.12.  
I'll try 0.1.13 later

**Comment by keisuke-nakata** [comment]
btw, what is the difference between 0.1.12 and 0.1.13?
It seems there is no update: https://github.com/minerllabs/minerl/compare/2e9079a79404e88283a2c2a69ed6b5d3b7758a82...75b5416eb97859971e3a8edbda670378fcbad709

**Comment by MadcowD** [comment]
This is worth trying again in 0.1.14 (I've fixed another race condition)

**Comment by MadcowD** [comment]
@keisuke-nakata Okay, so the fix isn't implemented here, but 
we need to effectively check to make sure that reset is reattempted when in the `reset` loop after `peek` is called from MineRL a `done=true` is returned.

**Comment by MadcowD** [comment]
@keisuke-nakata   This was fixed! Make sure to upgrade to `minerl` v`0.1.15`!
```
pip3 install --upgrade minerl
```
